### PR TITLE
Fix protobuf install command for Fedora Linux.

### LINKED ISSUE #6418: [Contributor Guide] Unclear Protobuf installation instructions
### Is your feature request related to a problem or challenge?

This [section](https://arrow.apache.org/datafusion/contributor-guide/index.html#protoc-installation) of the contributor guide explains how to install the Protobuf system dependencies:

> dnf install -y protobuf-compiler

This command is not sufficient to be able to build `arrow-datafusion`, at least not with the Fedora version I tested (Fedora 38).

### Describe the solution you'd like

Instead of suggesting `protobuf-compiler`, the docs should use `protobuf-devel` (which also installs the compiler).

> dnf install -y protobuf-devel

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by LouisGariepy**
# Which issue does this PR close?

Closes #6418.

# Rationale for this change

The current command `dnf install -y protobuf-compiler` does install a real package, but it is not sufficient to be able to build DataFusion on Fedora systems.

The proposed command `dnf install -y protobuf-devel` correctly installs the protobuf compiler as well as some development resources that are required.

The proposed change improves the docs in the sense that it installs the correct dependencies for Fedora

# What changes are included in this PR?

A small code snippet in the docs.

# Are these changes tested?

No tests required for such a documentation change, at least not as far as I'm aware.

# Are there any user-facing changes?

Yes, in the sense that it changes the documentation, but the current command simply doesn't work, and ultimately this change doesn't break anything for anyone.

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thank you @LouisGariepy

**Comment by tustvold** [comment]
> development resources that are required

What resources are these, I'm wondering if the actual issue is lack of tools like GCC necessary for building some crates, as opposed to anything protobuf specific?

**Comment by LouisGariepy** [comment]
**TLDR:** I definitely have `gcc` installed, I stumbled upon this problem while trying to build DataFusion on my main computer (which I use for daily Rust dev).

> You shouldn't need protobuf to build DataFusion

This seems to be incorrect (evidence below). As far as I can tell `cargo build` errors out saying that Protobuf is required, and in the case of Fedora Linux, the required dependencies are indeed acquired from `protobuf-devel`.

I provided the whole reproducer process below!

---

@tustvold Here's a reproducer for a more convincing argument. Sorry, it's a bit lengthy, but I wanted to make everything crystal clear:

# Reproducer
### Starting up a fresh Fedora container

```bash
podman run -it fedora bash
```

### Installing basic development libraries

```bash
# Linker
dnf install -y gcc
# VCS
dnf install -y git
```

### Installing Rust

```bash
curl https://sh.rustup.rs -sSf | sh -s -- -y
source "$HOME/.cargo/env"
```

### Checking that Rust is installed correctly with all the required dependencies

```bash
cargo new hello-world
cd hello-world
cargo run
cd ..
rm -rf hello-world
```

You should be greeted with a familiar sight:

```
     Created binary (application) `hello-world` package
   Compiling hello-world v0.1.0 (/hello-world)
    Finished dev [unoptimized + debuginfo] target(s) in 0.22s
     Running `target/debug/hello-world`
Hello, world!
```

This clearly shows that we are not missing any generic build dependencies. Rust is good to go!

### Clone and build DataFusion

```bash
# Let's pull in submodules for the sake of showing that we're not missing anything
git clone --recurse-submodules https://github.com/apache/arrow-datafusion.git
cd arrow-datafusion
cargo build
```

You will see this:
```
  thread 'main' panicked at 'Could not find `protoc` installation and this build crate cannot proceed without
      this knowledge. If `protoc` is installed and this crate had trouble finding
      it, you can set the `PROTOC` environment variable with the specific path to your
      installed `protoc` binary.If you're on debian, try `apt-get install protobuf-compiler` or download it from https://github.com/protocolbuffers/protobuf/releases
```

**Oh no**.

### Fixing the error by reading the docs
So, apparently Protobufs **are** required to build DataFusion. The savvy user will go straight to the contributor guide and see a reassuring line of code:

> `dnf install -y protobuf-compiler`

This should work, right? Let's try again:

```bash
cargo build
```

**Oh my!** We get a crazy cursed error message! Open the `details` collapsible section below to get a good scare.

<details>

Error: Custom { kind: Other, error: "protoc failed: google/protobuf/any.proto: File not found.\nsubstrait/extensions/extensions.proto:6:1: Import \"google/protobuf/any.proto\" was not found or had errors.\nsubstrait/extensions/extensions.proto:77:3: \"google.protobuf.Any\" is not defined.\nsubstrait/extensions/extensions.proto:80:3: \"google.protobuf.Any\" is not defined.\ngoogle/protobuf/empty.proto: File not found.\nsubstrait/type.proto:6:1: Import \"google/protobuf/empty.proto\" was not found or had errors.\nsubstrait/type.proto:192:7: \"google.protobuf.Empty\" is not defined.\nsubstrait/algebra.proto:6:1: Import \"google/protobuf/any.proto\" was not found or had errors.\nsubstrait/algebra.proto:7:1: Import \"substrait/extensions/extensions.proto\" was not found or had errors.\nsubstrait/algebra.proto:8:1: Import \"substrait/type.proto\" was not found or had errors.\nsubstrait/algebra.proto:46:7: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:52:7: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:40:5: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:25:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:78:5: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:89:5: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:140:9: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:95:5: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:60:3: \"NamedStruct\" is not defined.\nsubstrait/algebra.proto:64:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:152:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:178:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:187:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:198:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:214:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:236:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:244:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:254:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:271:3: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:277:3: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:284:3: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:338:7: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:303:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:380:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:386:3: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:397:3: \"NamedStruct\" is not defined.\nsubstrait/algebra.proto:447:3: \"NamedStruct\" is not defined.\nsubstrait/algebra.proto:510:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:537:3: \"substrait.extensions.AdvancedExtension\" is not defined.\nsubstrait/algebra.proto:544:5: \"Type\" is not defined.\nsubstrait/algebra.proto:696:16: \"Type.Parameter\" is not defined.\nsubstrait/algebra.proto:700:7: \"google.protobuf.Any\" is not defined.\nsubstrait/algebra.proto:627:7: \"Type\" is not defined.\nsubstrait/algebra.proto:629:7: \"Type.List\" is not defined.\nsubstrait/algebra.proto:630:7: \"Type.Map\" is not defined.\nsubstrait/algebra.proto:772:5: \"Type\" is not defined.\nsubstrait/algebra.proto:806:5: \"Type\" is not defined.\nsubstrait/algebra.proto:905:5: \"Type\" is not defined.\nsubstrait/algebra.proto:943:5: \"Type\" is not defined.\nsubstrait/algebra.proto:1258:3: \"Type\" is not defined.\n" }

</details>

### Here's where my suggestion comes in

Looking at the error message, we can see lines like

> `protoc failed: google/protobuf/any.proto: File not found`

Google maintains protobufs, and the error says we're missing some protobuf files, so my reasoning was that installing the development version of protobuf would solve this issue.

Let's remove the protobuf dependency we installed. This is not strictly required, but I want to showcase that my solution works without anything else being needed:

```bash
dnf remove -y protobuf-compiler
```

Now let's install the protobuf development version

```bash
dnf install -y protobuf-devel
```

And we try once again:
```bash
cargo build
```

Tadaa! :) Everything builds this time.

**Comment by tustvold** [comment]
Thank you for the comprehensive writeup, yes that makes sense. Strange that fedora doesn't include the standard protos with the compiler, but hey ho.

As for requiring protobuf, you are quite correct, as the workspace includes the binary for regenerating the protobufs you will need it to build the workspacd. The datafusion library crates themselves should not require it, a lot of effort has been made to ensure that 😅

As an aside perhaps we should remove the binary crate from the workspace to avoid that 🤔

**Comment by tustvold** [comment]
Thanks once again :)
